pkg/reconciler/tenancy/workspacemounts/workspacemounts_controller.go

```diff
@@ -197,20 +197,61 @@ func (c *Controller) process(ctx context.Context, key string) (bool, error) {
 	logger := logging.WithObject(klog.FromContext(ctx), workspace)
 	ctx = klog.NewContext(ctx, logger)
 
-	var errs []error
-	requeue, err := c.reconcile(ctx, workspace)
+	getMountObjectFunc := func(ctx context.Context, cluster logicalcluster.Path, ref tenancyv1alpha1.ObjectReference) (*unstructured.Unstructured, error) {
+		// TODO(sttts): do proper REST mapping.
+		resource := strings.ToLower(ref.Kind) + "s"
+		gvr := schema.GroupVersionResource{Resource: resource}
+		cs := strings.SplitN(ref.APIVersion, "/", 2)
+		if len(cs) == 2 {
+			gvr.Group = cs[0]
+			gvr.Version = cs[1]
+		} else {
+			gvr.Version = ref.APIVersion
+		}
+		return c.dynamicClusterClient.Cluster(cluster).Resource(gvr).Get(ctx, ref.Name, metav1.GetOptions{})
+	}
+
+	// reconcile the status
+	statusUpdater := &workspaceStatusUpdater{
+		getMountObject: getMountObjectFunc,
+	}
+
+	status, err := statusUpdater.reconcile(ctx, workspace)
 	if err != nil {
-		errs = append(errs, err)
+		return false, err
+	}
+
+	if status == reconcileStatusStopAndRequeue {
+		return true, nil
 	}
 
 	// If the object being reconciled changed as a result, update it.
 	oldResource := &workspaceResource{ObjectMeta: old.ObjectMeta, Spec: &old.Spec, Status: &old.Status}
-	newResource := &workspaceResource{ObjectMeta: workspace.ObjectMeta, Spec: &workspace.Spec, Status: &workspace.Status}
+	newResource := &workspaceResource{ObjectMeta: workspace.ObjectMeta, Spec: &old.Spec, Status: &workspace.Status}
+	if err := c.commit(ctx, oldResource, newResource); err != nil {
+		return false, err
+	}
+
+	// reconcole the spec
+	specUpdater := &workspaceSpecUpdater{
+		getMountObject: getMountObjectFunc,
+	}
+	status, err = specUpdater.reconcile(ctx, workspace)
```

Comment:
Can we get at least the comment why this breaks the standard patter? 
is this to avoid status and spec update in the same loop?

Reply:
Yes, the standard pattern is limited to updating only either spec or status.

I will add a comment to explain.

Reply:
Done.